### PR TITLE
🔁 712 Improve display of selected spans

### PR DESCRIPTION
**Pull request by xoscar**
This PR adds the next and prev buttons to the DAG view when there is at least one affected span in the store.

## Changes

- Next/Prev functionality.
- New slice for spans.
- Created hooks, services, and components to separate logic.

## Fixes

- https://github.com/kubeshop/tracetest/issues/712 (Partially, the search will be added on a separated PR)

## Checklist

- [x] tested locally
- [ ] added new dependencies
- [ ] updated the docs
- [x] added a test

https://www.loom.com/share/65459831e1414a3ba99632936ee2d4aa

### CODE REVIEW

File: web/src/components/AssertionForm/AssertionForm.tsx [hunk 1]
```diff
@@ -52,7 +52,7 @@ const AssertionForm: React.FC<TAssertionFormProps> = ({
   testId,
   runId,
 }) => {
-  const {setAffectedSpans} = useTestDefinition();
+  const {onSetAffectedSpans, onClearAffectedSpans} = useSpan();
```
Comment (xoscar) [inline web/src/components/AssertionForm/AssertionForm.tsx:55]: Now this is coming from the span provider

File: web/src/components/Diagram/components/DAG/Controls.tsx [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+interface IProps {
+  onSelectSpan(spanId: string): void;
+}
+
+const Controls = ({onSelectSpan}: IProps) => {
+  const {zoomIn, zoomOut} = useZoomPanHelper();
+  const {affectedSpans} = useSpan();
+  const {handleNextSpan, handlePrevSpan, indexOfFocused} = useControls({onSelectSpan});
+
+  return (
```
Comment (xoscar) [inline web/src/components/Diagram/components/DAG/Controls.tsx:16]: Next, prev and zoom buttons

File: web/src/components/Diagram/components/DAG/DAG.tsx [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+      spans.map(span => ({
+        id: span.id,
+        parentIds: span.parentId ? [span.parentId] : [],
+        data: span,
+        type: TraceNodes.TraceNode,
+      })),
+    [spans]
+  );
+
+  const elementList = useDAGChart<TSpan>(nodeList, selectedSpan, onSelectSpan);
```
Comment (xoscar) [inline web/src/components/Diagram/components/DAG/DAG.tsx:41]: Moved the dag logic to the hook

File: web/src/components/Diagram/components/DAG/hooks/useControls.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+const useControls = ({onSelectSpan}: IProps) => {
+  const {setCenter, fitView} = useZoomPanHelper();
+  const {focusedSpan, onSetFocusedSpan, affectedSpans} = useSpan();
+  const {getState} = useStore();
+  const indexOfFocused = affectedSpans.findIndex(spanId => spanId === focusedSpan) + 1;
+
+  const getNodePosition = useCallback(
+    (nodeId: string) => {
+      const {nodes} = getState();
+      const {position} = nodes.find(node => node.id === nodeId) || {};
```
Comment (xoscar) [inline web/src/components/Diagram/components/DAG/hooks/useControls.ts:18]: Gets the node position to move the dag view

File: web/src/providers/Span/Span.provider.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+});
+
+interface IProps {
+  children: React.ReactNode;
+}
+
+export const useSpan = () => useContext(Context);
+
+const SpanProvider = ({children}: IProps) => {
+  const dispatch = useAppDispatch();
```
Comment (xoscar) [inline web/src/providers/Span/Span.provider.tsx:35]: Consolidates the span slice actions

File: web/src/redux/slices/Span.slice.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    setFocusedSpan(state, {payload: {spanId}}: PayloadAction<{spanId: string}>) {
+      state.focusedSpan = spanId;
+    },
+    clearAffectedSpans(state) {
+      state.affectedSpans = [];
+      state.focusedSpan = '';
+    },
+  },
+  extraReducers: builder => {
+    builder.addCase(setSelectedAssertion, (state, {payload: assertionResult}) => {
```
Comment (xoscar) [inline web/src/redux/slices/Span.slice.ts:31]: side effect to grab the affected spans when an assertion is selected

File: web/src/services/DAG.service.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
+import {dagStratify, Dag, sugiyama, layeringSimplex, decrossOpt, coordCenter} from 'd3-dag';
```
Comment (xoscar) [inline web/src/services/DAG.service.ts:1]: Consolidates all of the DAG logic in one place

File: web/src/providers/Span/Span.provider.tsx [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+  const selectedSpan = useSelector(SpanSelectors.selectSelectedSpan);
+  const affectedSpans = useSelector(SpanSelectors.selectAffectedSpans);
+  const focusedSpan = useSelector(SpanSelectors.selectFocusedSpan);
+
+  const onSelectSpan = useCallback(
+    (span: TSpan) => {
```
Comment (cescoferraro) [inline web/src/providers/Span/Span.provider.tsx:41]: I know you guys disagree. 
But I would abstract each hook in here to a separate file, in order to unit test them individually

**Review by cescoferraro — APPROVED** [review]
Great work @xoscar !

**Review by jorgeepc — APPROVED** [review]
Looks good! 🔥

**Pull request merged**